tikv/raftstore/worker.go

```diff
@@ -624,11 +624,88 @@ func (r *regionRunner) shutdown() {
 	// todo, currently it is a a place holder.
 }
 
+type raftLogGcTaskRes uint64
+
 type raftLogGCRunner struct {
+	taskResCh chan<- raftLogGcTaskRes
+}
+
+// In our tests, we found that if the batch size is too large, running deleteAllInRange will
+// reduce OLTP QPS by 30% ~ 60%. We found that 32K is a proper choice.
+const MaxDeleteBatchSize int = 32 * 1024
+
+// gcRaftLog does the GC job and returns the count of logs collected.
+func (r *raftLogGCRunner) gcRaftLog(raftDb *badger.DB, regionId, startIdx, endIdx uint64) (uint64, error) {
+
+	// Find the raft log idx range needed to be gc.
+	firstIdx := startIdx
+	if firstIdx == 0 {
+		firstIdx = endIdx
+		err := raftDb.View(func(txn *badger.Txn) error {
+			startKey := RaftLogKey(regionId, 0)
+			ite := txn.NewIterator(badger.DefaultIteratorOptions)
+			defer ite.Close()
+			if ite.Seek(startKey); ite.Valid() {
+				var err error
+				if firstIdx, err = RaftLogIndex(ite.Item().KeyCopy(nil)); err != nil {
+					return err
+				}
+			}
+			return nil
+		})
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	if firstIdx >= endIdx {
+		log.Infof("no need to gc, [regionId: %d]", regionId)
+		return 0, nil
+	}
+
+	counter := uint64(0)
```

Comment:
Why use `counter` instead of `endIdx - firstIdx`?

Reply:
because error might happen and a partial raft log are gced during `writeToRaft`. I think it might be useful collecting raft log gc size even when error occurs, so by using counter, it can do this. For tikv, it just return `0` when error occurs.

Reply:
No, the transaction is atomic, either all succeed or all fail.